pandas/core/arrays/datetimelike.py

```diff
@@ -478,6 +478,52 @@ def __getitem__(self, key):
 
         return self._simple_new(result, **attribs)
 
+    def __setitem__(
+            self,
+            key,    # type: Union[int, Sequence[int], Sequence[bool], slice]
+            value,  # type: Union[NaTType, Scalar, Sequence[Scalar]]
+    ):
+        # type: (...) -> None
+        # I'm fudging the types a bit here. The "Scalar" above really depends
+        # on type(self). For PeriodArray, it's Period (or stuff coercible
+        # to a period in from_sequence). For DatetimeArray, it's Timestamp...
+        # I don't know if mypy can do that, possibly with Generics.
+        # https://mypy.readthedocs.io/en/latest/generics.html
+
+        if is_list_like(value):
+            is_slice = isinstance(key, slice)
+            if (not is_slice
+                    and len(key) != len(value)
```

Comment:
key could be a scalar here right? (in which case u will get an odd exception about len of unsized object)

Reply:
Hrm... yeah

```python
In [3]: pd.period_range('2000', periods=2)._data
Out[3]:
<PeriodArray>
['2000-01-01', '2000-01-02']
Length: 2, dtype: period[D]

In [4]: arr = pd.period_range('2000', periods=2)._data

In [5]: arr[0] = arr[[0, 1]]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-9a9ffe90069d> in <module>
----> 1 arr[0] = arr[[0, 1]]

~/sandbox/pandas/pandas/core/arrays/datetimelike.py in __setitem__(self, key, value)
    494             is_slice = isinstance(key, slice)
    495             if (not is_slice
--> 496                     and len(key) != len(value)
    497                     and not com.is_bool_indexer(key)):
    498                 msg = ("shape mismatch: value array of length '{}' does not "

TypeError: object of type 'int' has no len()
```

This is a gap in the base extension tests as well. I'll ad one. that should be a... what? ValueError for trying to set list-like into a single element?

Reply:
NumPy raises with `ValueError: setting an array element with a sequence.` which seems fine to me.

Reply:
yep seems good
maybe just let it fall thru this if will work?

Reply:
Seemed a tad easier to explicitly check for a scalar `key` here.

Actually, grr, this is kinda annoying but NumPy allows setting a sequence of length 1.

```python
In [2]: x = np.array([1, 2])

In [3]: x[0] = x[[0]]
```

but now I would raise on that.

```python
In [10]: arr = pd.period_range('2000', periods=2)._data

In [11]: arr[0] = arr[[0]]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-666653b9c59a> in <module>
----> 1 arr[0] = arr[[0]]

~/sandbox/pandas/pandas/core/arrays/datetimelike.py in __setitem__(self, key, value)
    495
    496             if lib.is_scalar(key):
--> 497                 raise ValueError("setting an array element with a sequence.")
    498
    499             if (not is_slice

ValueError: setting an array element with a sequence.
```

Thoughts? I think this should raise, since a length-1 sequence is more like a sequence than a scalar.

Reply:
yeah i think we raise on this

this is why setitem in Block is so complicated :)